### PR TITLE
Refactor DatabaseManagement

### LINKED ISSUE #126: Refactor `DatabaseManagement`
needs to be database-agnostic, arguably composable

other problems:
- catching `IllegalArgumentException` is way too broad: we could potentially snuff unrelated exception. We need domain-specific ways to signal error conditions.

### PR DESCRIPTION
**Pull request by bbjubjub2494**
Closes #126

depends on #101

### CODE REVIEW

**Review by wattsecs — APPROVED** [review]
Very necessary refactoring. One small question but otherwise all good in my opinion

File: app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatabaseManagementTest.kt [hunk 1]
```diff
@@ -347,24 +347,6 @@ class FirestoreDatabaseManagementTest {
         assertThat(demoManagement.getDatasetIds().first(), equalTo("PzuR0B48GpYN5ERxM3DW"))
     }
 
-    @Test
```
[the remark this answers is not in the document]
Comment (bbjubjub2494) [reply app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatabaseManagementTest.kt:350]: The behavior (allowing operation on deleted/fake datasets) is not supported by the design under test. I moved it to `ConcurrentDeleteDatabaseManagement`

File: app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/firestore/OfflineDatabaseManagementTest.kt [hunk 2]
```diff
@@ -372,24 +358,6 @@ class OfflineDatabaseManagementTest {
         MatcherAssert.assertThat(demoManagement.getDatasetIds().first(), Matchers.equalTo("PzuR0B48GpYN5ERxM3DW"))
     }
 
-    @Test
```
[the remark this answers is not in the document]
Comment (bbjubjub2494) [reply app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/firestore/OfflineDatabaseManagementTest.kt:375]: ditto

**Review by martin-lenweiter — APPROVED** [review]
Good changes. Improves code readability and modularity. The concurrent delete implementation is a nice addition, too.

**Pull request merged**